[MSTransferor] Create rule for each pileup location

amaltaro opened the pull request:
Fixes #10975 

#### Status
ready

#### Description
The initial objective with this PR was the following:
* enforce the campaign pileup configuration, such that each location defined there gets a rucio rule for the whole container (grouping=ALL).
* secondary AAA can trigger secondary input data location as well, in case a location defined in the campaign does not have a rule locking it
* for workflows with multiple pileup datasets, use an intersection of their location defined in the campaign configuration for the primary/parent data placement
* as before, the pileup location defines the workflow sitelist (for primary and parent data placement). If secondary is enabled, then primary/parent can go to other locations as defined in the sitelist.
* special handling for RelVal, which do not define secondaries in the campaign, in that case simply use the workflow sitelist as candidate location for the pileup dataset(s)

However, I decided to refactor many things that were in place to deal with the old PhEDEx-based data placement. A summary of those changes are:
* no longer increment RSE usage within our own cache, instead only rely on what Rucio provides
* no longer create chunks of primary/parent input blocks, simply place all the blocks with grouping=DATASET against a logical OR of the final RSEs
* DQMHarvest workflows will keep getting their input blocks locked with grouping=ALL against a logical OR of the final RSEs (we no longer pick 1 RSE for that)
* workflows with OpenRunningTimeout will get a rule locking the whole input dataset with grouping=DATASET against a logical OR of the final RSEs
* workflows with primary and parent dataset will get a rule locking all the input blocks together, thus grouping=ALL against a logical OR of the final RSEs
* completely remove the logic to find the best RSE, let Rucio take care of that.

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
Given that it modifies a recent feature introduced with: https://github.com/dmwm/WMCore/pull/11141
it needs to be properly tested again.

#### External dependencies / deployment changes
Service configuration disabling verbose logs:
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/147
and
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/148

amaltaro committed:
[MSTransferor] Create rule for each pileup location

fix data returned from getChunkBlocks method

do not make it a list of a set

special case for secondaries in RelVals

another bugfix for the secondary relval case

remove confusing log about location used

amaltaro committed:
unit test updates

more unit tests

unit tests with the Workflow class relocation

bunch of new unit tests for the new templates

update unit tests with the getChunkBlocks removal

remove getChunkBlocks from unit tests as well

amaltaro commented:
MSTransferor level tests are looking okay, but I am letting those workflows to go through the system to see how it goes until the very end. 

I think review can start though. @haozturk could you please have a look at this as well. I tried to highlight all the important changes in the initial description, but if you see any wrong/missing use case, please let me know.

vkuznet reviewed (commented):
I need to admit that review of such PR requires detailed knowledge of data management system which I don't have. Therefore, I can't comment on why so much stuff has been removed and then added to MSTransferor. Moreover, it seems that the entire structure of MSTransferor requires code-refactoring based on provided data types. Right now, the code handle everything via if/else structures, while I would expect that code itself should be very abstract, i.e. here is data and here are the rules (methods) to call over the data. While, concrete implementations of different data-types should be separated. If you'll have a classes for "Neutrino", "DQMHarvesting", "primary", etc. (whatever data-type data management will need to handle) and let these classes implement common methods, then it would be much more easier to maintain the code. In this design, the MSTransferor will only ask input data `what is your type`, and then call appropriate class methods which is associated with such data type. This will keep MSTransferor logic abstract and generic, while also will allow to implement different rules/methods for different types of data. Said that, I don't have concrete comments to the proposed changes as I, honestly, do not understand the logic of data transfers.

Here is how I envision `decideDataPlacement ` method should looks like:
```
def __init__(self):
       self.trasferorObjects = {
            'Neutrino': NeutrinoTransferor()
             'DQMHarvester': DQMHarvestTransferor()
              ....
       }
def decideDataPlacement(self, wflow, dataIn):
       dataType, groupping = self.getDataType(wflow)
       self.trasnferorObject[dataType].decideDataPlacement(wflow, datain, groupping)

def getDataType(self, wflow):
       if wflow.getReqType() == "DQMHarvest":
           return "DQMHarvest", "ALL"
       if wflow.getOpenRunningTimeout() > self.msConfig["openRunning"]:
           return "OpenRunning", "DATASET"
```
And, then within this MS area (`WMCore/MicroServices/MSTransferor`) you'll add `DQMHarvestTransferor.py`, `NeutrinoTransferor.py`, etc. modules which will keep implementation for concrete use-cases. This will make code more maintainable and extensible since when new use-case will pop-up you'll only need to add new implementation. And, if logic for specific use-case will require changes you'll only change this class logic, while MSTransferor logic will be very generic.

amaltaro commented:
Valentin, thanks for these comments! I will have to think whether this implementation is feasible and how to do so.

On what concerns the large code removal, I mentioned this in the initial PR review. But it was possible because much of the MSTransferor logic was still based on how we used to work with PhEDEx; while now we can delegate much of the data management logic to the actual data management system :)

todor-ivanov reviewed (commented):
Thanks @amaltaro  for those huge changes here.  In general the implementation looks good. Besides the few minor comments I have left inline I am having the impression that:
*  With those changes a big portion of the previously automated logic i s now completely dropped, and the data placement would rely completely on the campaign level configuration
* The input data placement becomes much more static than it was before. 
* We loosen the constraints on the service side on what goes where much more than before

There are a lot of changes that  go in with the current PR, which are not completely related to reorganizing the secondary location logic, but rather to code refactoring. And I know the best  moment to work on something like that is when you are actually reiterating through the code, but I think it could be good to have a separate issue mentioning that refactoring, which to be resolved with the current PR as well. But if you think that would be too much of effort just skip this request of mine.

We would also benefit from a well described current policy of the service behavior as it is right now, so that we can update the relevant section in the documentation [here](https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Transferor). Which I believe we need to review together with P&R Team.

amaltaro committed:
Reorganization and classification of workflows and policies

complete code reorganization

fix getInputSecData and GrowingWorkflow

remove dual space from RequestInfo

remove useless init method from sub-classes

remove duplicate Workflow instantiation in RequestInfo

apply Todor and Valentin suggestions

change a few log level line from debug to info

remove unused import

amaltaro commented:
@vkuznet Valentin, could you please have a quick look at my 3rd commit. It's very much raw, but I just wanted to make it available to see whether that is more or less what you were suggesting above? If so, then I can proceed with this. With that code in mind, each class would behave slightly different and also provide the final setup for Rucio rules. There are still a few cases where we have all sort of data in the same workflow, and in that case it's hard to classify anything!

Thanks for your review, Todor. I apologize again for providing so much changes with something that was supposed to be small, but I thought the small changes would only make things more complex and ugly, reason why I made this substantial refactoring.
Yes, with these changes, things are no longer much in our control and we start depending more on the data management system to manage data (which is ideal IMO). But we had some clarification on this over zoom as well.
I will look and work on your comments as soon as we decide how to proceed with these changes (another reorganization might be coming up).

vkuznet reviewed (commented):
Yes, I think it is proper architecture, and each Workflow has a set of common methods. Then, individual details will be delegated to specific classes. I only found one issue with return and what kind of return it should be.

amaltaro commented:
Thanks for the confirmation, I just wanted to make sure I was on the right direction. I will work on the remaining implementation for Monday or so.

todor-ivanov commented on src/python/WMCore/MicroService/MSTransferor/DataStructs/DQMHarvestWorkflow.py:
```diff
@@ -0,0 +1,35 @@
+
+        :return: a string with the required DID grouping
+        """
+        return "ALL"
```
wouldn't it be better to have this one configurable?

amaltaro replied:
It would for the MSTransferor. But at this object layer - and given that there are different objects with different attributes/values - I think having it as a constant is a good compromise without making this code more complex.

If I refactor the code and move it outside of the Workflow object, then it's much more doable. Let me give it further thought.

todor-ivanov commented on src/python/WMCore/MicroService/MSTransferor/DataStructs/GrowingWorkflow.py:
```diff
@@ -0,0 +1,37 @@
+
+        :return: a string with the required DID grouping
+        """
+        return "DATASET"
```
same here?

todor-ivanov commented on src/python/WMCore/MicroService/MSTransferor/DataStructs/RelValWorkflow.py:
```diff
@@ -0,0 +1,33 @@
+
+        :return: a string with the required DID grouping
+        """
+        return "ALL"
```
same here

todor-ivanov commented on src/python/WMCore/MicroService/MSTransferor/DataStructs/Workflow.py:
```diff
@@ -484,3 +484,25 @@ def getOpenRunningTimeout(self):
+        and transferred with Rucio
+        :return: a list of unique blocks
+        """
+        return self.getChunkBlocks()
```
why do you need a method for doing a single return of a call to another method. I can understand that, though, if it is needed for other methods naming compatibility and following the same structure.

todor-ivanov commented on src/python/WMCore/MicroService/MSTransferor/DataStructs/Workflow.py:
```diff
@@ -484,3 +484,25 @@ def getOpenRunningTimeout(self):
+        Returns the rucio rule grouping to be defined for in primary
+        and/or parent input data placement, where:
+            * ALL: all blocks are placed under the same RSE
+            * DATASET: blocks can be scattered in multiple RSEs
```
Please make a note here stating that we are talking about Rucio `DATASET` rather than a CMS `DATASET`

todor-ivanov commented on src/python/WMCore/MicroService/MSTransferor/DataStructs/Workflow.py:
```diff
@@ -484,3 +484,25 @@ def getOpenRunningTimeout(self):
+
+        :return: a string with the required DID grouping
+        """
+        return "DATASET"
```
Again, even for the default workflows the grouping policy may be good to have it configurable, such that future request for changes in system behavior are addressed only through service configurations.

todor-ivanov commented on src/python/WMCore/MicroService/MSTransferor/RequestInfo.py:
```diff
@@ -72,6 +70,33 @@ def __call__(self, reqRecords):
 
+    def classifyWorkflows(self, reqRecords):
+        """
+        This method classifies input workflows, making it easier
```
What does "input workflow" mean in this context? My feeling is that this method classifies all workflows.

todor-ivanov reviewed (approved):
Hi @amaltaro I've made few minor comments inline. None of them is a show stopper. To me the code looks good enough to be merged (provided you think there is nothing more to be added as well).

vkuznet commented on src/python/WMCore/MicroService/MSTransferor/DataStructs/Workflow.py:
```diff
@@ -484,3 +484,25 @@ def getOpenRunningTimeout(self):
         """
         return self.data.get("OpenRunningTimeout", 0)
+
+    def getInputData(self):
```
You don't need this method implementation here since it is defined in `Workflow` class and since it is identical it can be simply dropped. This method will be take from a parent class.

vkuznet commented on src/python/WMCore/MicroService/MSTransferor/MSTransferor.py:
```diff
@@ -352,221 +330,80 @@ def checkDataLocation(self, wflow):
+        else:
+            # then there are multiple pileups, get an intersection of their
+            # expected location as final workflow destination
+            wflowFinalLocation = campBasedLocation[0].intersection(*campBasedLocation)
```
I'm not sure I understand logic of operation here, i.e. you are taking first element of `campBasedLocation` and intersect it with the entire list. Why first? I think you should explain the operation in a comment above by clarifying what is stored in a first element and why intersection with entire list is required, e.g. the first element of `campBasedLocation ` is original pileup, the rest are additional pileups and need to intersect them together.

vkuznet reviewed (changes requested):
In addition to Todor's comments about "ALL", "DATASET", etc. I suggest to define separate module (or class) which will keep all grouping definitions over there as constants. Then, you can use these constants over here and avoid hard-coded string patterns scattered through the code. When new grouping will be required you'll define new constant in such module/class and use it.

amaltaro commented on src/python/WMCore/MicroService/MSTransferor/DataStructs/GrowingWorkflow.py:
```diff
@@ -0,0 +1,37 @@
+        :return: a list of unique blocks
+        """
+        inputContainer = [self.getInputDataset()]
+        currentTotalSize = sum([blockInfo['blockSize'] for blockInfo in self.getPrimaryBlocks().values()])
```
I wanted to distinguish it from `totalBlockSize` because it actually is the total size at this very moment. As this input dataset will keep growing (new blocks will appear), I wanted to make it more explicit with the variable name.
But `totalBlockSize` isn't a bad idea for consistency.

amaltaro commented on src/python/WMCore/MicroService/MSTransferor/MSTransferor.py:
```diff
@@ -352,221 +330,80 @@ def checkDataLocation(self, wflow):
-        :return: a string with the final pileup destination PNN
-        """
-        # FIXME: workflows should be marked as failed if there is no common
-        # site between SiteWhitelist and secondary location
```
They are not marked as failed, but there is a Prometheus alert. This way people can react and get it through the system.

amaltaro commented on src/python/WMCore/MicroService/MSTransferor/MSTransferor.py:
```diff
@@ -352,221 +330,80 @@ def checkDataLocation(self, wflow):
+                    campSecLocations = campConfig['Secondaries'].get(dsetName, [])
+                campBasedLocation.append(set(campSecLocations))
+
+                if not campSecLocations:
```
This is how it was before, but now we have N replicas of the pileup, where N is the number of locations in the campaign. Before, we had a single replica enforced (but from discussions with Hasan, they were **manually** creating replicas for all the other locations. So, I'd say we are at the same state, just with more automation now :-D

amaltaro commented:
@todor-ivanov @vkuznet I think I addressed most of your comments, other than making the grouping setting configurable. I'd rather leave that one to the future (if it's really needed). The last 2 commits have those recent changes.

todor-ivanov reviewed (approved):
Thanks @amaltaro 
It looks good to me.

amaltaro commented:
unit tests are failing with while contacting DBS, e.g.:
```
DBS Server error: [{'error': {'reason': 'dbs error', 'message': '', 'function': 'dbs.parameters.CheckQueryParameters', 'code': 108}, 'http': {'method': 'GET', 'code': 400, 'timestamp': '2022-05-18 16:19:10.348970569 +0000 UTC m=+1411.120187744', 'path': '/dbs/prod/global/DBSReader/datasets?dataset=/MinimumBias/ComissioningHI-v1/RAW&dataset_access_type=*&detail=False', 'user_agent': 'DBSClient/Unknown/', 'x_forwarded_host': 'dbs-prod.cern.ch', 'x_forwarded_for': '137.138.157.32', 'remote_addr': '137.138.63.204:44858'}, 'exception': 400, 'type': 'HTTPError', 'message': 'DBSError Code:108 Description:DBS file load error, e.g. fail to load DB template Function:dbs.parameters.CheckQueryParameters Message: Error: dbs error'}]
```

@vkuznet I know you are working on a new release, ping'ing you just so you are aware of this.

amaltaro commented:
test this please

vkuznet commented:
sorry, I didn't update configuration file (doing it now). Things should be back to normal now.

vkuznet commented:
test this please

amaltaro commented:
Thank you very much for the review and great suggestions, Valentin and Todor.
